File: sdks/typescript/src/domains.ts
import { isHostCovered, normalizeHost } from "./hosts.js";
import type { DomainEntry } from "./models.js";

/**
 * An immutable snapshot of `GET /v1/domains`, with the host matching built in.
 *
 * The endpoint is `Cache-Control: max-age=300` behind a strong `ETag` and
 * clients are expected to revalidate rather than refetch, so a snapshot carries
 * the `etag` that produced it. Pass it back to
 * `OctogenClient#fetchDomainCoverage` and a `304` returns this same object.
 */
export class DomainCoverage {
  /** Every `(host, catalog)` pair, exactly as the server returned it. */
  readonly entries: readonly DomainEntry[];
  /** The strong `ETag` of this snapshot; revalidate with it. */
  readonly etag: string | undefined;
  /** `max-age` from `Cache-Control`, in seconds, when the server sent one. */
  readonly maxAgeSeconds: number | undefined;
  /** When this snapshot was received, for `maxAgeSeconds` staleness checks. */
  readonly retrievedAt: Date;

  private readonly catalogsByHost: ReadonlyMap<string, readonly string[]>;

  constructor(
    entries: readonly DomainEntry[],
    options: {
      etag?: string | undefined;
      maxAgeSeconds?: number | undefined;
      retrievedAt?: Date | undefined;
    } = {},
  ) {
    this.entries = Object.freeze([...entries]);
    this.etag = options.etag;
    this.maxAgeSeconds = options.maxAgeSeconds;
    this.retrievedAt = options.retrievedAt ?? new Date();

    const catalogsByHost = new Map<string, string[]>();
    for (const entry of this.entries) {
      // Normalize the server's side too. Today it only emits apex hosts, but a
      // `www.` entry appears in the published example, and normalizing both
      // sides means such an entry still matches rather than becoming a
      // duplicate key that nothing looks up.
      const host = normalizeHost(entry.host);
      if (host === undefined) {
        continue;
      }
      const catalogs = catalogsByHost.get(host);
      if (catalogs === undefined) {
        catalogsByHost.set(host, [entry.catalog]);
      } else if (!catalogs.includes(entry.catalog)) {
        catalogs.push(entry.catalog);
      }
    }
    this.catalogsByHost = catalogsByHost;
  }

  /** The normalized covered hosts — the raw set, for callers that want it. */
  get hosts(): readonly string[] {
    return [...this.catalogsByHost.keys()].sort();
  }

  /**
   * Is this URL's merchant covered?
   *
   * Takes a full product URL or a bare host and normalizes it before
   * comparing, so `https://www.macys.com/shop/product/x` matches the
   * `macys.com` the server reports.
   */
  isHostCovered(urlOrHost: string | null | undefined): boolean {
    const host = normalizeHost(urlOrHost);
    return host !== undefined && this.catalogsByHost.has(host);
  }

  /** Catalogs claiming this URL's host; empty when it is not covered. */
  catalogsFor(urlOrHost: string | null | undefined): readonly string[] {
    const host = normalizeHost(urlOrHost);
    if (host === undefined) {
      return [];
    }
    return this.catalogsByHost.get(host) ?? [];
  }

  /** True once `max-age` has elapsed — time to revalidate with `etag`. */
  isStale(now: Date = new Date()): boolean {
    if (this.maxAgeSeconds === undefined) {
      return true;
    }
    const ageSeconds = (now.getTime() - this.retrievedAt.getTime()) / 1000;
    return ageSeconds >= this.maxAgeSeconds;
  }
}

export { isHostCovered, normalizeHost };

File: sdks/python/src/octogen_ai_sdk/domains.py
"""Covered-domain snapshot for ``GET /v1/domains``."""

from __future__ import annotations

from collections.abc import Iterable, Sequence
from datetime import UTC, datetime

from octogen_ai_sdk.generated.models import DomainEntry
from octogen_ai_sdk.hosts import normalize_host


class DomainCoverage:
    """An immutable snapshot of ``GET /v1/domains``, with host matching built in.

    The endpoint is ``Cache-Control: max-age=300`` behind a strong ``ETag`` and
    clients are expected to revalidate rather than refetch, so a snapshot
    carries the ``etag`` that produced it. Pass it back to
    :meth:`OctogenClient.fetch_domain_coverage` and a ``304`` returns this same
    object.
    """

    __slots__ = (
        "_catalogs_by_host",
        "entries",
        "etag",
        "max_age_seconds",
        "retrieved_at",
    )

    def __init__(
        self,
        entries: Sequence[DomainEntry],
        *,
        etag: str | None = None,
        max_age_seconds: int | None = None,
        retrieved_at: datetime | None = None,
    ) -> None:
        self.entries: tuple[DomainEntry, ...] = tuple(entries)
        self.etag = etag
        self.max_age_seconds = max_age_seconds
        self.retrieved_at = retrieved_at or datetime.now(UTC)

        catalogs_by_host: dict[str, list[str]] = {}
        for entry in self.entries:
            # Normalize the server's side too. Today it only emits apex hosts,
            # but a `www.` entry appears in the published example; normalizing
            # both sides means such an entry still matches instead of becoming
            # a duplicate key nothing looks up.
            host = normalize_host(entry.host)
            if host is None:
                continue
            catalogs = catalogs_by_host.setdefault(host, [])
            if entry.catalog not in catalogs:
                catalogs.append(entry.catalog)
        self._catalogs_by_host = catalogs_by_host

    @property
    def hosts(self) -> tuple[str, ...]:
        """The normalized covered hosts — the raw set, for callers who want it."""
        return tuple(sorted(self._catalogs_by_host))

    def is_host_covered(self, url_or_host: str | None) -> bool:
        """Is this URL's merchant covered?

        Takes a full product URL or a bare host and normalizes it before
        comparing, so ``https://www.macys.com/shop/product/x`` matches the
        ``macys.com`` the server reports.
        """
        host = normalize_host(url_or_host)
        return host is not None and host in self._catalogs_by_host

    def catalogs_for(self, url_or_host: str | None) -> tuple[str, ...]:
        """Catalogs claiming this URL's host; empty when it is not covered."""
        host = normalize_host(url_or_host)
        if host is None:
            return ()
        return tuple(self._catalogs_by_host.get(host, ()))

    def is_stale(self, now: datetime | None = None) -> bool:
        """True once ``max-age`` has elapsed — revalidate with :attr:`etag`."""
        if self.max_age_seconds is None:
            return True
        moment = now or datetime.now(UTC)
        age = (moment - self.retrieved_at).total_seconds()
        return age >= self.max_age_seconds

    def __repr__(self) -> str:
        return (
            f"DomainCoverage(hosts={len(self._catalogs_by_host)}, "
            f"entries={len(self.entries)}, etag={self.etag!r})"
        )


def coverage_hosts(entries: Iterable[DomainEntry]) -> tuple[str, ...]:
    """Normalized hosts from raw ``/v1/domains`` entries, deduped and sorted."""
    return DomainCoverage(list(entries)).hosts
